targets/psa_switch/psa_switch.cpp

```diff
@@ -366,46 +483,24 @@ PsaSwitch::ingress_thread() {
     const auto &f_ig_cos = phv->get_field("psa_ingress_output_metadata.class_of_service");
     const auto ig_cos = f_ig_cos.get_uint();
 
-    // handling multicast
-    unsigned int mgid = 0u;
-    const auto &f_mgid = phv->get_field("psa_ingress_output_metadata.multicast_group");
-    mgid = f_mgid.get_uint();
-
+    unsigned int mgid = phv->get_field("psa_ingress_output_metadata.multicast_group").get_uint();
     if (mgid != 0) {
       BMLOG_DEBUG_PKT(*packet,
                       "Multicast requested for packet with multicast group {}",
                       mgid);
-      const auto pre_out = pre->replicate({mgid});
-      auto &f_instance = phv->get_field("psa_egress_input_metadata.instance");
-      auto &f_eg_cos = phv->get_field("psa_egress_input_metadata.class_of_service");
-      auto packet_size = packet->get_register(PACKET_LENGTH_REG_IDX);
-      for(const auto &out : pre_out){
-        auto egress_port = out.egress_port;
-        auto instance = out.rid;
-        BMLOG_DEBUG_PKT(*packet,
-                        "Replicating packet on port {} with instance {}",
-                        egress_port, instance);
-        f_instance.set(instance);
-        // TODO use appropriate enum member from JSON
-        f_packet_path.set(PACKET_PATH_NORMAL_MULTICAST);
-        f_eg_cos.set(ig_cos);
-        std::unique_ptr<Packet> packet_copy = packet->clone_with_phv_ptr();
-        packet_copy->set_register(PACKET_LENGTH_REG_IDX, packet_size);
-        enqueue(egress_port, std::move(packet_copy));
-      }
+      multicast(packet.get(), mgid, PACKET_PATH_NORMAL_MULTICAST);
```

Comment:
Do you know whether these changes are based upon the latest version of the behavioral-model code?   I know I have (unintentionally) made your life a little more difficult here by making changes to this code recently that have been merged in while your work was in progress.

Your proposed multicast function does not include the assignments to the class_of_service metadata field, for example, and there is now a PSA test for normal multicasting of packets (not involving cloning) that will fail if class_of_service is not copied from ingress to egress for such packets, as required by the PSA specification.

For cloned packets, the class_of_service should come from the clone session configuration, but I do not know if that is something that is part of the clone session configuration in your changes right now.  It could be left for a later enhancement if that is easier, preferably with a comment that it should be done later.

Perhaps you could add a new parameter 'egress_cos' to your multicast function, and pass in the value 'ig_cos' when called from here for normal multicast packets, and put in a constant 0 for now when you call multicast from the ingress-to-egress clone code that does multicast?  That would be the place for the TBD comment to change the 0 later.

Reply:
I accidentally started a review instead of commenting on this, then removed the review and thereby my response to this question.

I added in your suggested class_of_service handling to multicast() + a TBD comment.

Reply:
Yep, I saw that, and looks good to me.

As I mentioned in another comment, it is important not only that with these changes, p4lang/behavioral-model automated tests on p4lang/behavioral-model repository pass, which is automatically initiated and recorded by Github when you commit here, but there is currently nothing automated to tell whether these changes will cause p4lang/p4c automated tests to pass or fail, until AFTER the p4lang/behavioral-model changes are merged in, and a few hours pass, and the next p4lang/p4c automated tests run for one of its PR commits.

If you know how to locally run the p4lang/p4c tests with these changes locally in your p4lang/behavioral-model repository, please let us know if those all pass (again, perhaps with ebpf and ubpf tests failing, but that is OK if you haven't done the necessary steps p4lang/p4c install steps for those to work, which I do not currently know what they are).